yarn.lock

```diff
@@ -1310,6 +1332,11 @@
   dependencies:
     "@codemirror/text" "^0.19.0"
 
+"@codemirror/state@^6.0.0", "@codemirror/state@^6.1.2":
```

Comment:
This is troubling — it means the app has multiple copies of codemirror state/view (and other) packages. This breaks in a lots of ways (if there's global state it won't work, if there's `instanceof` check it won't work and etc). Probably this is why syntax highlighting doesn't work.

Reply:
How would we best fix this?

Reply:
Like this: https://classic.yarnpkg.com/lang/en/docs/cli/install/#toc-yarn-install-flat?

Reply:
lang-bqn requires codemirror@6 while the app — codemirror@0.19

so the proper way to fix would be to update the app to use codemirror@6 which is (a) not difficult (mostly mechanical, as there are just renames, see [the diff](https://github.com/mechanize-systems/bqnpad/commit/4c8ab77f79a576b02cbe46f2af912b6f9b63d99e) for lang-bqn) but (b) annoying as you have to make sure all language integrations you use are up to date as well (hopefully they are)

(amusingly I just recently upgraded lang-bqn before publishing it as I thought the ecosystem has moved on)

Reply:
I'm not sure we can update to codemirror 6 as SimpleMDE relies on it being CodeMirror 5. We have both CodeMirror 5 and CodeMirror 6 running fine together.

Before we spend time on trying to solve this, can I understand why this plugin requires things like `codemirror/autocomplete` and `codemirror/state`? How are those being used to do syntax highlighting? Thanks.

Reply:
> I'm not sure we can update to codemirror 6 as SimpleMDE relies on it being CodeMirror 5. We have both CodeMirror 5 and CodeMirror 6 running fine together.

Sorry I wasn't clear enough — I was talking about codemirror6@0.19 to codemirror6@6 upgrade, so only codemirror6 is affected, codemirror5 can stay as-is.

> Before we spend time on trying to solve this, can I understand why this plugin requires things like codemirror/autocomplete and codemirror/state? How are those being used to do syntax highlighting? Thanks.

`lang-bqn` provides not just syntax highlighting but also BQN glyph input method and completion for glyphs. You can see how it works at [BQNPAD](http://bqnpad.mechanize.systems).

If you want to pull just the syntax highlighting it's possible to use  `lezer-bqn` package and [some piece of code](https://github.com/mechanize-systems/bqnpad/blob/main/lang-bqn/index.ts#L18-L70) from `lang-bqn` to make it work. But... `lezer-bqn` requires a recent lezer version as well (lezer is a parsing infra used by codemirror6 for syntax highlighting and other means) and it works with recent codemirror6 version...

Reply:
For a quick&dirty workaround, I think, it makes sense to recover the pre-codemirror/lezer upgrade version of lang-bqn/lezer-bqn by taking code from BQNPAD directly before [this commit](https://github.com/mechanize-systems/bqnpad/commit/4c8ab77f79a576b02cbe46f2af912b6f9b63d99e).

Reply:
> Sorry I wasn't clear enough — I was talking about codemirror6@0.19 to codemirror6@6 upgrade, so only codemirror6 is affected, codemirror5 can stay as-is.

I see. Then yes, that feels feasible to do, as long as it doesn't break all the other syntax highlighters we have. Thanks for explaining!

So @ErikSchierboom maybe you could try that, then if it causes trouble, go with @andreypopp's quick&dirty suggestion? 🙂

Reply:
If I examine the output of `yarn why @codemirror/state`, I get the following:

```shell
erik@schierheim:~/exercism/website$ yarn why @codemirror/state
yarn why v1.22.19
[1/4] Why do we have the module "@codemirror/state"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@codemirror/state@0.19.6"
info Has been hoisted to "@codemirror/state"
info Reasons this module exists
   - Hoisted from "@codemirror#language#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#state"
   - Hoisted from "@codemirror#lang-javascript#@codemirror#state"
   - Hoisted from "@codemirror#stream-parser#@codemirror#state"
   - Hoisted from "@codemirror#theme-one-dark#@codemirror#state"
   - Hoisted from "lang-julia#@codemirror#state"
   - Hoisted from "@codemirror#view#@codemirror#state"
   - Hoisted from "@codemirror#autocomplete#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#closebrackets#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#comment#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#fold#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#gutter#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#matchbrackets#@codemirror#state"
   - Hoisted from "@codemirror#view#@codemirror#rangeset#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#rectangular-selection#@codemirror#state"
   - Hoisted from "@codemirror#autocomplete#@codemirror#tooltip#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#commands#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#history#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#highlight#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#search#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#lint#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#lint#@codemirror#panel#@codemirror#state"
info Disk size without dependencies: "248KB"
info Disk size with unique dependencies: "344KB"
info Disk size with transitive dependencies: "344KB"
info Number of shared dependencies: 1
=> Found "lang-bqn#@codemirror/state@6.1.2"
info Reasons this module exists
   - "lang-bqn" depends on it
   - Hoisted from "lang-bqn#@codemirror#autocomplete#@codemirror#state"
   - Hoisted from "lang-bqn#@codemirror#language#@codemirror#state"
   - Hoisted from "lang-bqn#@codemirror#view#@codemirror#state"
info Disk size without dependencies: "388KB"
info Disk size with unique dependencies: "388KB"
info Disk size with transitive dependencies: "388KB"
info Number of shared dependencies: 0
Done in 0.30s.
```

To me that looks like upgrading would potentially break a lot of things. `yarn upgrade --latest @codemirror/state` also doesn't change anything.

Reply:
I've tried pinning the package to a commit: 

```json
"lang-bqn": "https://github.com/mechanize-systems/bqnpad#4842c490a5830b7482bd7d59ddbfb15c22bb676a",
```

However, that still uses a different version:

```shell
erik@schierheim:~/exercism/website$ yarn why @codemirror/state
yarn why v1.22.19
[1/4] Why do we have the module "@codemirror/state"...?
[2/4] Initialising dependency graph...
[3/4] Finding dependency...
[4/4] Calculating file sizes...
=> Found "@codemirror/state@0.19.6"
info Has been hoisted to "@codemirror/state"
info Reasons this module exists
   - Specified in "dependencies"
   - Hoisted from "@codemirror#basic-setup#@codemirror#state"
   - Hoisted from "@codemirror#lang-javascript#@codemirror#state"
   - Hoisted from "@codemirror#stream-parser#@codemirror#state"
   - Hoisted from "@codemirror#theme-one-dark#@codemirror#state"
   - Hoisted from "lang-julia#@codemirror#state"
   - Hoisted from "@codemirror#view#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#closebrackets#@codemirror#state"
   - Hoisted from "lang-bqn#@codemirror#collab#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#comment#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#fold#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#gutter#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#language#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#matchbrackets#@codemirror#state"
   - Hoisted from "@codemirror#view#@codemirror#rangeset#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#rectangular-selection#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#commands#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#history#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#highlight#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#search#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#autocomplete#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#lint#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#lint#@codemirror#panel#@codemirror#state"
   - Hoisted from "@codemirror#basic-setup#@codemirror#autocomplete#@codemirror#tooltip#@codemirror#state"
info Disk size without dependencies: "248KB"
info Disk size with unique dependencies: "344KB"
info Disk size with transitive dependencies: "344KB"
info Number of shared dependencies: 1
=> Found "lang-bqn#@codemirror/state@0.19.9"
info This module exists because "lang-bqn" depends on it.
info Disk size without dependencies: "248KB"
info Disk size with unique dependencies: "344KB"
info Disk size with transitive dependencies: "344KB"
info Number of shared dependencies: 1
Done in 0.32s.
```

And even more importantly, it doesn't compile:

```shell
js       | ✘ [ERROR] Could not resolve "lang-bqn"
js       |
js       |     app/javascript/components/misc/CodeMirror/languageCompartment.ts:245:35:
js       |       245 │       const { bqn } = await import('lang-bqn')
js       |           ╵                                    ~~~~~~~~~~
js       |
js       |   You can mark the path "lang-bqn" as external to exclude it from the bundle, which will remove this error. You can also add ".catch()" here to handle this failure at run-time instead of bundle-time.
```

Reply:
And I'm then also getting this in the console log:

```shell
chunk-P6ML2NA7.js:2556 Uncaught (in promise) Error: Unrecognized extension value in extension set ([object Object]). This sometimes happens because multiple instances of @codemirror/state are loaded, breaking instanceof checks.
```

Reply:
> To me that looks like upgrading would potentially break a lot of things.

Most of the things listed are official codemirror packages which I'm sure are up to date themselves (regarding dependency on @codemirror/state) and also have only minor changes since 0.19 (0.19 -> 0.20 is described [here](https://discuss.codemirror.net/t/release-0-20-0/4302) and then 0.20 -> 6.0 has some minor renames, described [here](https://discuss.codemirror.net/t/codemirror-6-0-has-been-released/4498)).

The only package up to date is `lang-julia` (I'm co-maintainer there, so can help releasing it updated).

> I've tried pinning the package to a commit:

The repo hosts multiple packages, you cannot pin a single package like that from there (don't know if yarn supports doing that at all).

Reply:
@andreypopp But just upgrading the codemirror dependencies doesn't work, we still get different version of the state component _and_ our code stops working. I've tried upgrading to codemirror 6, but that made all sorts of things break. I'm gonna stop making all this work, as it is costing me too much time to do so.

> For a quick&dirty workaround, I think, it makes sense to recover the pre-codemirror/lezer upgrade version of lang-bqn/lezer-bqn by taking code from BQNPAD directly before [this commit](https://github.com/mechanize-systems/bqnpad/commit/4c8ab77f79a576b02cbe46f2af912b6f9b63d99e).

Please provide instructions on how to do this.

Reply:
> Please provide instructions on how to do this.

you can just vendor code into excercism/website repo (take it before the commit I've linked to), you need both lezer-bqn and lang-bqn

Reply:
Not to be annoying, but couldn't you create an NPM package for that commit?

Reply:
I can package, sure, but it'd be nice if you can confirm first the referenced code works with exorcism first (can be done by just vendoring code inside the app).

Reply:
@andreypopp Ah like that. How do I create the `bqn.grammar.js` file that is referenced in https://github.com/mechanize-systems/bqnpad/blob/4842c490a5830b7482bd7d59ddbfb15c22bb676a/lezer-bqn/index.js?

Reply:
Little bump on this :)

Reply:
@ErikSchierboom need to use `lezer-generator`, the invocation in the [Makefile](https://github.com/mechanize-systems/bqnpad/blob/4842c490a5830b7482bd7d59ddbfb15c22bb676a/Makefile#L41)

Reply:
@andreypopp I get:

> Uncaught RangeError: Parser version (14) doesn't match runtime version (13)